### ISSUE TITLE
RedisPipelineException-WriteRedisConnectionException-invalid commands

### ISSUE THREAD

**Issue by syschen** [redisson/redisson#2649]
报错信息：
Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is org.springframework.data.redis.connection.RedisPipelineException: Pipeline contained one or more invalid commands; nested exception is org.redisson.client.WriteRedisConnectionException: Unable to write command into connection! Node source: NodeSource [slot=null, addr=null, redisClient=null, redirect=null, entry=MasterSlaveEntry [masterEntry=[freeSubscribeConnectionsAmount=0, freeSubscribeConnectionsCounter=value:50:queue:0, freeConnectionsAmount=31, freeConnectionsCounter=value:63:queue:0, freezed=false, freezeReason=null, client=[addr=redis://redis-m.xxx.com:6379], nodeType=MASTER, firstFail=0]]], connection: RedisConnection@1628911 [redisClient=[addr=redis://redis-m.xxx.com:6379], channel=[id: 0xed6e09bd, L:/1xx.sss.xxx.xxx:sss - R:redis-m.xxx.com/xxx.xx.x6.2s02:sss], currentCommand=null], command: null, params: null after 3 retry attempts] with root cause 
java.lang.NullPointerException: null

配置：
masterSlaveServersConfig:
  idleConnectionTimeout: 10000
  connectTimeout: 10000
  timeout: 3000
  retryAttempts: 3
  retryInterval: 1500
  failedSlaveReconnectionInterval: 3000
  failedSlaveCheckInterval: 60000
  password: null
  subscriptionsPerConnection: 5
  clientName: null
  loadBalancer: !<org.redisson.connection.balancer.RoundRobinLoadBalancer> {}
  subscriptionConnectionMinimumIdleSize: 1
  subscriptionConnectionPoolSize: 50
  slaveConnectionMinimumIdleSize: 32
  slaveConnectionPoolSize: 64
  masterConnectionMinimumIdleSize: 32
  masterConnectionPoolSize: 64
  readMode: "SLAVE"
  subscriptionMode: "SLAVE"
  slaveAddresses:
#    - "redis://127.0.0.1:6381"
    - "redis://redis-s.xxxx.com:6379"
  masterAddress: "redis://redis-m.xxxx.com:6379"
  database: 0
threads: 0
nettyThreads: 0
codec: !<org.redisson.codec.FstCodec> {}
transportMode: "NIO"

调用场景：
用户退出登录时，删除缓存中的token和用户信息
consumerTokenServices.revokeToken(logoutDto.getToken());
tokenCacheService.removeCellphoneTokenKey();
tokenCacheService.removeCellphoneToken(tokenCacheKey);

报错信息里面提示了currentCommand=null，但是没有了更细的日志了
不知是什么原因导致command=null呢

谢谢！

**Comment by mrniko** [comment]
> with root cause
java.lang.NullPointerException: null

Could you share full stacktrace?

**Comment by syschen** [comment]
> > with root cause
> > java.lang.NullPointerException: null
> 
> Could you share full stacktrace?



2020-03-23 14:14:30 | N/A | 10.244.202.17 | 8 | ada-oauth | [redisson-timer-4-1] | | DEBUG | org.redisson.command.RedisExecutor | run | 239 | - attempt 1 for command null and params null
2020-03-23 14:14:32 | N/A | 10.244.202.17 | 8 | ada-oauth | [redisson-timer-4-1] | | DEBUG | org.redisson.command.RedisExecutor | lambda$releaseConnection$7 | 634 | - connection released for command null and params null from slot NodeSource [slot=null, addr=null, redisClient=null, redirect=null, entry=MasterSlaveEntry [masterEntry=[freeSubscribeConnectionsAmount=0, freeSubscribeConnectionsCounter=value:50:queue:0, freeConnectionsAmount=32, freeConnectionsCounter=value:64:queue:0, freezed=false, freezeReason=null, client=[addr=redis://redis-m.adachinainternal.com:6379], nodeType=MASTER, firstFail=0]]] using connection RedisConnection@2037605269 [redisClient=[addr=redis://redis-m.adachinainternal.com:6379], channel=[id: 0x1dba3c4e, L:/:38572 - R:redis-m.adachinainternal.com/6379], currentCommand=null]
2020-03-23 14:14:32 | N/A | 10.244.202.17 | 8 | ada-oauth | [redisson-timer-4-1] | | DEBUG | org.redisson.command.RedisExecutor | run | 239 | - attempt 2 for command null and params null
2020-03-23 14:14:33 | N/A | 10.244.202.17 | 8 | ada-oauth | [redisson-timer-4-1] | | DEBUG | org.redisson.command.RedisExecutor | lambda$releaseConnection$7 | 634 | - connection released for command null and params null from slot NodeSource [slot=null, addr=null, redisClient=null, redirect=null, entry=MasterSlaveEntry [masterEntry=[freeSubscribeConnectionsAmount=0, freeSubscribeConnectionsCounter=value:50:queue:0, freeConnectionsAmount=32, freeConnectionsCounter=value:64:queue:0, freezed=false, freezeReason=null, client=[addr=redis://redis-m.adachinainternal.com:6379], nodeType=MASTER, firstFail=0]]] using connection RedisConnection@1649883814 [redisClient=[addr=redis://redis-m.adachinainternal.com:6379], channel=[id: 0xd80a8043, L:/:38510 - R:redis-m.adachinainternal.com/:6379], currentCommand=null]
2020-03-23 14:14:33 | N/A | 10.244.202.17 | 8 | ada-oauth | [redisson-timer-4-1] | | DEBUG | org.redisson.command.RedisExecutor | run | 239 | - attempt 3 for command null and params null
2020-03-23 14:14:33 | N/A | 10.244.202.17 | 8 | ada-oauth | [redisson-netty-2-4] | | DEBUG | org.redisson.command.RedisExecutor | lambda$releaseConnection$7 | 634 | - connection released for command null and params null from slot NodeSource [slot=null, addr=null, redisClient=null, redirect=null, entry=MasterSlaveEntry [masterEntry=[freeSubscribeConnectionsAmount=0, freeSubscribeConnectionsCounter=value:50:queue:0, freeConnectionsAmount=32, freeConnectionsCounter=value:64:queue:0, freezed=false, freezeReason=null, client=[addr=redis://redis-m.adachinainternal.com:6379], nodeType=MASTER, firstFail=0]]] using connection RedisConnection@1597584999 [redisClient=[addr=redis://redis-m.adachinainternal.com:6379], channel=[id: 0xd2c556a2, L:/38514 - R:redis-m.adachinainternal.com/:6379], currentCommand=null]
2020-03-23 14:14:33 | 1309.47.15849440691829667 | 10.244.202.17 | 8 | ada-oauth | [http-nio-8848-exec-3] | | ERROR | o.a.c.c.C.[Tomcat].[localhost].[/].[dispatcherServlet] | log | 175 | - Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is org.springframework.data.redis.connection.RedisPipelineException: Pipeline contained one or more invalid commands; nested exception is org.redisson.client.WriteRedisConnectionException: Unable to write command into connection! Node source: NodeSource [slot=null, addr=null, redisClient=null, redirect=null, entry=MasterSlaveEntry [masterEntry=[freeSubscribeConnectionsAmount=0, freeSubscribeConnectionsCounter=value:50:queue:0, freeConnectionsAmount=31, freeConnectionsCounter=value:63:queue:0, freezed=false, freezeReason=null, client=[addr=redis://redis-m.adachinainternal.com:6379], nodeType=MASTER, firstFail=0]]], connection: RedisConnection@1597584999 [redisClient=[addr=redis://redis-m.adachinainternal.com:6379], channel=[id: 0xd2c556a2, L:/:38514 - R:redis-m.adachinainternal.com/:6379], currentCommand=null], command: null, params: null after 3 retry **attempts] with root cause
java.lang.NullPointerException: null**
2020-03-23 14:14:34 | 1309.74.15837273194760001 | 10.244.202.17 | 8 | ada-oauth | [parallel-2] | | DEBUG | org.redisson.command.RedisExecutor | sendCommand | 612 | - acquired connection for command (INFO DEFAULT) and params [] from slot NodeSource [slot=0, addr=null, redisClient=null, redirect=null, entry=MasterSlaveEntry [masterEntry=[freeSubscribeConnectionsAmount=0, freeSubscribeConnectionsCounter=value:50:queue:0, freeConnectionsAmount=32, freeConnectionsCounter=value:64:queue:0, freezed=false, freezeReason=null, client=[addr=redis://redis-m.adachinainternal.com:6379], nodeType=MASTER, firstFail=0]]] using node redis-s.adachinainternal.com/:6379... RedisConnection@517103590 [redisClient=[addr=redis://redis-s.adachinainternal.com:6379], channel=[id: 0xcdcde602, L:/1:38490 - R:redis-s.adachinainternal.com/:6379], currentCommand=null]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
sorry no further detail could be found
seems relate to  Redis pipeline 
code may related as below:

![image](https://user-images.githubusercontent.com/8849212/77289001-093d3080-6d14-11ea-8ad4-06c62340bd50.png)

**Comment by mrniko** [comment]
Which Redisson version do you use?

**Comment by syschen** [comment]
> Which Redisson version do you use?

3.11.5

**Comment by mrniko** [comment]
Could you share code example of RBatch you use?

**Comment by syschen** [comment]
> Could you share code example of RBatch you use?

We use method spring-security-oauth2:2.3.3.RELEASE/RedisTokenStore/removeAccessToken
as our LOGOUT service.The code may cause PipelineException as below:

public void removeAccessToken(String tokenValue) {
		byte[] accessKey = serializeKey(ACCESS + tokenValue);
		byte[] authKey = serializeKey(AUTH + tokenValue);
		byte[] accessToRefreshKey = serializeKey(ACCESS_TO_REFRESH + tokenValue);
		RedisConnection conn = getConnection();
		try {
			**conn.openPipeline();**
			conn.get(accessKey);
			conn.get(authKey);
			conn.del(accessKey);
			conn.del(accessToRefreshKey);
........................

**Comment by syschen** [comment]
@Override
    public List<Object> closePipeline() throws RedisPipelineException {
        if (isPipelined()) {
            CommandBatchService es = (CommandBatchService) executorService;
            try {
                BatchResult<?> result = es.execute();
                filterResults(result);
                if (isPipelinedAtomic()) {
                    return Arrays.<Object>asList((List<Object>) result.getResponses());
                }
                return (List<Object>) result.getResponses();
            } catch (Exception ex) {
                **throw new RedisPipelineException(ex);**
            } finally {
                resetConnection();
            }
        }
        return Collections.emptyList();
    }

Is there any chance  this method-**_closePipeline_**-may cause  NPE?

**Comment by mrniko** [comment]
Unable to reproduce it with 3.12.3 version. I think it's related to this https://github.com/redisson/redisson/issues/2419

**Comment by mrniko** [comment]
Fixed in https://github.com/redisson/redisson/issues/2680

**Comment by syschen** [comment]
thanks for update


 原始邮件 
发件人: Nikita Koksharov<notifications@github.com>
收件人: redisson/redisson<redisson@noreply.github.com>
抄送: syschen<syschen@126.com>; Author<author@noreply.github.com>
发送时间: 2020年4月9日(周四) 19:49
主题: Re: [redisson/redisson]RedisPipelineException-WriteRedisConnectionException-invalid commands (#2649)


Fixed in #2680
